docbrown/core/src/lazy_vec.rs

```diff
@@ -0,0 +1,269 @@
+    // values from "Props::prop_ids" hashmap
+    LazyVec1(usize, A),
+    LazyVecN(Vec<A>),
+}
+
+impl<A> LazyVec<A>
+where
+    A: PartialEq + Default + Clone
+{
+    pub(crate) fn from(id: usize, value: A) -> Self {
+        LazyVec::LazyVec1(id, value)
+    }
+
+    pub(crate) fn filled_ids(&self) -> Vec<usize> {
+        match self {
+            LazyVec::Empty => Default::default(),
+            LazyVec::LazyVec1(id, _) => vec![*id],
+            LazyVec::LazyVecN(vector) => {
+                vector.iter().enumerate()
+                    .filter(|&(id, value)| *value != Default::default())
+                    .map(|(id, _)| id)
+                    .collect()
+            }
+        }
+    }
+
+    pub(crate) fn get_mut(&mut self, id: usize) -> Option<&mut A> {
+        match self {
+            LazyVec::LazyVec1(only_id, value) if *only_id == id => Some(value),
+            LazyVec::LazyVecN(vec) => vec.get_mut(id),
+            _ => None,
+        }
+    }
+
+    pub(crate) fn get(&self, id: usize) -> Option<&A> {
+        match self {
+            LazyVec::LazyVec1(only_id, value) if *only_id == id => Some(value),
+            LazyVec::LazyVecN(vec) => vec.get(id),
+            _ => None,
+        }
+    }
+
+    // fails if there is already a value set for the given id and it's different from the default
+    pub(crate) fn set(&mut self, id: usize, value: A) -> Result<(), ()> {
+        match self {
+            LazyVec::Empty => {
+                *self = Self::from(id, value);
+                Ok(())
+            }
+            LazyVec::LazyVec1(only_id, only_value) => {
+                if *only_id == id {
+                    Err(())
+                } else {
+                    let mut vector = vec![Default::default(); usize::max(id, *only_id) + 1];
+                    vector[id] = value;
+                    vector[*only_id] = only_value.clone();
+                    *self = LazyVec::LazyVecN(vector);
+                    Ok(())
+                }
+            }
+            LazyVec::LazyVecN(vector) => {
+                if vector.len() <= id {
+                    vector.resize(id + 1, Default::default())
+                }
+                if vector[id] == Default::default() {
+                    vector[id] = value;
+                    Ok(())
+                } else {
+                    Err(())
+                }
+            }
+        }
+    }
+
+    pub(crate) fn update_or_set<F>(&mut self, id: usize, mut updater: F, default: A)
+    where
+        F: FnMut(&mut A) // TODO: is FnMut right?
+    {
+        match self.get_mut(id) {
+            Some(value) => updater(value),
+            None => self.set(id, default).unwrap(),
+        }
+    }
+
+
+
+    fn handle_illegal_change(id: usize) {
+        panic!("Tried to change meta data for id '{}'", id); // TODO: move this somewhere else and use it for static props
+    }
+}
+
+// TODO: move all these test to Props ?
```

Comment:
Think we can del

Reply:
the tests were conceptually the same as the ones in tprop.rs and I added some new tests for this new implementation